File: pattern-searching/boyermoore/boyermoore.java
package boyermoore;

public class boyermoore {
    public static final int NO_OF_CHARS = 256;

    public static int max(int a, int b) {
        return (a > b) ? a : b;
    }

    public static void badCharHeuristic(char[] str, int size, int badchar[]) {
        int i;

        // Initialize all occurences as -1
        for (i = 0; i < NO_OF_CHARS; i++) {
            badchar[i] = -1;
        }

        // Fill the actual value of last occurences of a character
        for (i = 0; i < size; i++) {
            badchar[(int) str[i]] = i;
        }
    }

    // A pattern searching function that uses Bad Character Heuristic of Bouer Moore
    // Algorithm
    public static void search(char txt[], char pat[]) {
        int m = pat.length;
        int n = txt.length;
        int[] badchar = new int[NO_OF_CHARS];

        /*
         * Fill the bad character array by calling the preprocessing function
         * badCharHeuristic for given pattern
         */
        badCharHeuristic(pat, m, badchar);

        int s = 0; // s is shift of the pattern respect to text
        while (s <= (n - m)) {
            int j = m - 1;
            /*
             * keep reducing index j of pattern while characters of pattern and text
             * arematching at this shift s
             */
            while (j >= 0 && pat[j] == txt[s + j])
                j--;

            /*
             * If the pattern is present at current shift, then index j will become -1 after
             * the aboove loop
             */
            if (j < 0) {
                System.out.println(s);
                s += (s + m < n) ? m - badchar[txt[s + m]] : 1;
            } else {
                s += max(1, j - badchar[txt[s + j]]);
            }
        }
    }

    public static void main(String[] args) {
        char txt[] = "ABAAABCD".toCharArray();
        char pat[] = "ABC".toCharArray();
        search(txt, pat);
    }
}

File: pattern-searching/boyermoore/boyermoore.py
NO_OF_CHARS = 256


def badCharHeuristic(str, size):
    global NO_OF_CHARS
    i = 0

    # Initialize all occurences as -1
    badchar = [-1]*NO_OF_CHARS

    # fill the actual value of last occurences of a character
    for i in range(0, size):
        badchar[ord(str[i])] = i

    return badchar


def search(txt, pat):
    global NO_OF_CHARS
    m = len(pat)
    n = len(txt)

    # Fill the bad characters by calling the preprocessing function
    # badCharHeuristic for gien pattern
    badchar = badCharHeuristic(pat, m)
    s = 0
    while(s <= (n-m)):
        j = m-1
        # keep reducing index j of pattern while characters of pattern and text
        # are matching at this shift s
        while(j >= 0 and pat[j] == txt[s+j]):
            j -= 1

        # If the pattern is present at current shift, then index j will become -1 after the above loop
        if(j < 0):
            print(s)
            s += (m - badchar[ord(txt[s+m])] if s+m<n else 1)
        else:
            s += max(1, j-badchar[ord(txt[s+j])])

if __name__ == "__main__":
    txt="ABAAABCD"
    pat="ABC"
    search(txt, pat)
